timer.h

```diff
@@ -9,36 +9,120 @@
 #include "win32/time.h"
 #include "win32/unistd.h"
 #include "win32/getrusage.h"
+#elif __APPLE__
+#include <mach/mach_init.h>
+#include <mach/mach_time.h>
 #else
-
-#include <sys/time.h>
+#include "sys/time.h"
 #include <sys/resource.h>
 #endif
 
-#if __APPLE__
-#include <mach/mach_init.h>
-#include <mach/mach_time.h>
+static int determine_clock_source(int clock_use_rdtsc) {
+#if defined(__APPLE__)
+    return TIDEWAYS_XHPROF_CLOCK_MACH;
+#elif defined(__powerpc__) || defined(__ppc__)
+    return TIDEWAYS_XHPROF_CLOCK_TSC;
+#elif defined(__s390__) // Covers both s390 and s390x.
+    return TIDEWAYS_XHPROF_CLOCK_TSC;
+#elif defined(__ARM_ARCH)
+    return TIDEWAYS_XHPROF_CLOCK_GTOD;
+#elif defined(PHP_WIN32)
+    return TIDEWAYS_XHPROF_CLOCK_QPC;
+#else
+    struct timespec res;
+
+    if (clock_use_rdtsc == 1) {
+        return TIDEWAYS_XHPROF_CLOCK_TSC;
+    }
+
+#if HAVE_CLOCK_GETTIME
+    if (clock_gettime(CLOCK_MONOTONIC, &res) == 0) {
+        return TIDEWAYS_XHPROF_CLOCK_CGT;
+    }
 #endif
 
+#if HAVE_GETTIMEOFDAY
+    return TIDEWAYS_XHPROF_CLOCK_GTOD;
+#endif
+
+    return TIDEWAYS_XHPROF_CLOCK_TSC;
+#endif
+}
+
 static zend_always_inline uint64 current_timestamp() {
     struct timeval tv;
 
     if (gettimeofday(&tv, NULL)) {
         php_error(E_ERROR, "tracer: Cannot acquire gettimeofday");
-        return 0;
+        zend_bailout();
     }
 
     return 1000 * (uint64) tv.tv_sec + (uint64) tv.tv_usec / 1000;
 }
 
+static zend_always_inline uint64 time_milliseconds_cgt()
+{
+#if HAVE_CLOCK_GETTIME
+    struct timespec s;
+
+    if (clock_gettime(CLOCK_MONOTONIC, &s) == 0) {
+        return s.tv_sec * 1000000 + s.tv_nsec / 1000;
+    }
+#endif
+
+    return 0;
+}
+
+static zend_always_inline uint64 time_milliseconds_gtod()
+{
+#if HAVE_GETTIMEOFDAY
+    struct timeval now;
+    if (gettimeofday(&now, NULL) == 0) {
+        return now.tv_sec * 1000000 + now.tv_usec;
+    }
+#endif
+
+    return 0;
+}
+
+static zend_always_inline uint64 time_milliseconds_tsc_query()
+{
+#if defined(__s390__) // Covers both s390 and s390x.
+
+    uint64_t tsc;
+    // Return the CPU clock.
+    asm("stck %0" : "=Q" (tsc) : : "cc");
+
+    return tsc;
+#elif defined(__i386__)
+    int64_t ret;
+    __asm__ volatile("rdtsc" : "=A"(ret));
+    return ret;
+#elif defined(__x86_64__) || defined(__amd64__)
+    uint64 val;
+    uint32 a, d;
+    asm volatile("rdtsc" : "=a" (a), "=d" (d));
+    (val) = ((uint64)a) | (((uint64)d)<<32);
+    return val;
+#elif defined(__powerpc__) || defined(__ppc__)
+    uint64 val;
+    asm volatile ("mftb %0" : "=r" (val));
+    return val;
+#else
+    return 0;
```

Comment:
Does it mean that arm will use that result?

Reply:
Yes, ARM automatically uses GTOD by default and doesn't support TSC for now.

See https://github.com/tideways/php-xhprof-extension/pull/74/files#diff-9c31fd9ea16faaddb9f8d2d6a605e19bR28